activesupport/lib/active_support/message_verifier.rb

```diff
@@ -46,16 +46,17 @@ def initialize(secret, options = {})
     def valid_message?(signed_message)
       return if signed_message.nil? || !signed_message.valid_encoding? || signed_message.blank?
 
-      data, digest = signed_message.split("--")
-      data.present? && digest.present? && ActiveSupport::SecurityUtils.secure_compare(digest, generate_digest(data))
+      parts = signed_message.split('.')
+      parts.size == 3 && parts.all?(&:present?) && untampered?(parts.pop, parts.join('.'))
     end
 
-    # Decodes the signed message using the +MessageVerifier+'s secret.
+    # Decodes the signed message using the +MessageVerifier+'s secret and returns the claims hash.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
     #
-    #   signed_message = verifier.generate 'a private message'
-    #   verifier.verified(signed_message) # => 'a private message'
+    #   signed_message = verifier.generate('a private message', for: 'test', expires_at: 1.day.from_now)
+    #   verifier.verified(signed_message)
+    #   # => {:pld=>"a private message", :for=>"test", :exp=>"2015-06-29T07:16:12.579Z"}
```

Comment:
I'm sorry you've been caught off guard by this. Weren't we clear enough about wanting to maintain backwards compatibility? Because wanting backwards compatibility automatically means we have to consider what methods return (and more).

As an aside, I want to talk more generally about how you respond to feedback. I feel like it's been happening a lot that you respond extremely confidently, but you almost don't ask any questions. Let's take this comment thread as an example. You've just seen something that you didn't know you had to be aware of and you have no questions about that at all? There's really no other things you think should be cleared up before you start working on this? I will be honest and say that worries me a bit :smile:

I guess my problem is that almost never hear what you're **thinking**, you just do stuff. Then I tell you to change something and you change it. I would rather, we could learn to anticipate each other, so we didn't need so many cycles to get somewhere. Does that make sense? :smile: